File: src/lib/functions/shoppingListTools.ts

import { FunctionDeclaration, SchemaType } from "@google/generative-ai";

export const showShoppingListTool: FunctionDeclaration = {
    name: "showShoppingList",
    description: "Shows the shopping list UI popup to the user. To get the content of the list, use getShoppingList.",
    parameters: {
      type: SchemaType.OBJECT,
      properties: {},
    },
};

export const getShoppingListTool: FunctionDeclaration = {
  name: "getShoppingList",
  description: "Retrieves and reads out the user's current shopping list.",
  parameters: {
    type: SchemaType.OBJECT,
    properties: {},
  },
};

export const addToShoppingListTool: FunctionDeclaration = {
  name: "addToShoppingList",
  description: "Adds one or more items to the user's shopping list. Use this function when the user confirms they want to add items you have suggested.",
  parameters: {
    type: SchemaType.OBJECT,
    properties: {
      items: {
        type: SchemaType.ARRAY,
        description: "A list of items to add to the shopping list.",
        items: {
          type: SchemaType.OBJECT,
          properties: {
            item: { type: SchemaType.STRING, description: "The name of the item (e.g., 'Milk', 'Bread')." },
            quantity: { type: SchemaType.NUMBER, description: "The quantity of the item." },
            unit: { type: SchemaType.STRING, description: "The unit of measurement (e.g., 'gallon', 'loaf', 'piece')." },
          },
          required: ["item", "quantity", "unit"]
        }
      }
    },
    required: ["items"],
  },
};

export const removeFromShoppingListTool: FunctionDeclaration = {
  name: "removeFromShoppingList",
  description: "Removes one or more items from the user's shopping list, for example when they have been purchased.",
  parameters: {
    type: SchemaType.OBJECT,
    properties: {
      item_names: {
        type: SchemaType.ARRAY,
        description: "A list of item names to remove from the shopping list.",
        items: {
          type: SchemaType.STRING,
        }
      }
    },
    required: ["item_names"],
  },
};

File: mise-asi/registry/schemas/shopping_list_tools.py
"""
Shopping list tool schemas
Maps to: src/lib/functions/shoppingListTools.ts and crudShoppingListTools.ts
"""

# Legacy tools - matching shoppingListTools.ts
show_shopping_list_tool = {
    "name": "showShoppingList",
    "description": "Opens the shopping list panel in the UI to show users their current shopping list. Use when user wants to see their shopping list.",
    "input_schema": {
        "type": "object",
        "properties": {},
        "required": []
    }
}

get_shopping_list_tool = {
    "name": "getShoppingList",
    "description": "Retrieves the current shopping list items. Use to check what's on the list before making suggestions.",
    "input_schema": {
        "type": "object",
        "properties": {},
        "required": []
    }
}

add_to_shopping_list_tool = {
    "name": "addToShoppingList",
    "description": "Adds items to the user's shopping list.",
    "input_schema": {
        "type": "object",
        "properties": {
            "items": {
                "type": "array",
                "description": "Items to add to the shopping list",
                "items": {
                    "type": "object",
                    "properties": {
                        "item": {"type": "string", "description": "Name of the item"},
                        "quantity": {"type": "number", "description": "Quantity needed"},
                        "unit": {"type": "string", "description": "Unit of measurement"}
                    },
                    "required": ["item"]
                }
            }
        },
        "required": ["items"]
    }
}

remove_from_shopping_list_tool = {
    "name": "removeFromShoppingList",
    "description": "Removes items from the shopping list by name.",
    "input_schema": {
        "type": "object",
        "properties": {
            "item_names": {
                "type": "array",
                "items": {"type": "string"},
                "description": "Names of items to remove"
            }
        },
        "required": ["item_names"]
    }
}

# CRUD tools - matching crudShoppingListTools.ts
get_shopping_list_items_tool = {
    "name": "getShoppingListItems",
    "description": "Get all shopping list items for the current user.",
    "input_schema": {
        "type": "object",
        "properties": {},
        "required": []
    }
}

create_shopping_list_items_tool = {
    "name": "createShoppingListItems",
    "description": "Create one or more new shopping list items.",
    "input_schema": {
        "type": "object",
        "properties": {
            "items": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {
                        "item": {"type": "string"},
                        "quantity": {"type": "number"},
                        "unit": {"type": "string"}
                    },
                    "required": ["item"]
                }
            }
        },
        "required": ["items"]
    }
}

delete_shopping_list_items_tool = {
    "name": "deleteShoppingListItems",
    "description": "Delete shopping list items by name.",
    "input_schema": {
        "type": "object",
        "properties": {
            "item_names": {
                "type": "array",
                "items": {"type": "string"},
                "description": "Names of items to delete"
            }
        },
        "required": ["item_names"]
    }
}
